[CSApply] Fix an issue with the @objc selector fix-it incorrectly being applied

theblixguy opened the pull request:
This PR fixes an issue where the `@objc` fix-it would be applied to the method declaration in the protocol. This would happen if you provided a default implementation of the method via a protocol extension and tried using it as a selector inside another method within the same extension. The fix-it would then be applied to the method on the protocol.

```
protocol Foo {
  func bar()
  func baz() // The `@objc` fix-it would then be applied here (incorrect).
}

extension Foo {
  func bar() {

   // This would trigger an error diagnostic that would offer the fix-it.
    let sel = #selector(baz)
  }

  func baz() {
    print("Hello, world)")
  }
}
```

With this fix, no fix-it will be applied as it might not actually work and might not be what the user expects.

```
protocol Foo {
  func bar()
  func baz()
}

extension Foo {
  func bar() {

    // This would trigger an error diagnostic. No fix-it will be offered.
    let sel = #selector(baz)
  }

  func baz() {
    print("Hello, world)")
  }
}
```

Resolves [SR-9391](https://bugs.swift.org/browse/SR-9391).

Commit:
[csapply] Fix an issue where objc selector fixit would be applied to the method on the protocol

theblixguy committed:
[csapply] fix indentation

theblixguy commented:
cc @harlanhaskins @jrose-apple @slavapestov @jckarter

theblixguy committed:
[csapply] fix indentation

Commit:
[csapply] fix indent again

CodaFi commented:
Could you please add a test.

CodaFi commented on lib/Sema/CSApply.cpp:
```diff
@@ -4196,12 +4196,26 @@ namespace {
+	// FIXME: Put the fix-it on the method declaration in the protocol
+	// extension, instead of the protocol itself?
+	auto methodContextDecl = foundDecl->getDeclContext()->getAsDecl();
+	bool isMethodDeclaredInProtocol = methodContextDecl->getKind() == DeclKind::Protocol;
```
You can use `foundDecl->getDeclContext()->getExtendedProtocolDecl()` here.

theblixguy replied:
Actually, `foundDecl->getDeclContext()->getExtendedProtocolDecl()` returns null, so I think I need to use `getSelfProtocolDecl()` instead.

CodaFi commented:
There's a really narrow case where this still won't be the right fixit 

```
protocol Foo {
  associatedtype Uhoh
  func bar()
  func baz()
}
```

If you try to mark `Foo` `@objc` you're in the same boat.

Commit:
[csapply] gardening

Commit:
[csapply] use getExtendedProtocolDecl

Commit:
[csapply] use getSelfProtocolDecl() instead and remove redundant declkind check

jrose-apple commented:
I don't think we want to suggest making a Swift protocol `@objc` when it wasn't before. That's a drastic change in many cases, and it doesn't solve the default implementation problem. Just not offering the fix-it is probably good enough, and if you wanted to go further you could add a note like "members of protocol extensions cannot be exposed to Objective-C".

theblixguy commented:
@jrose-apple Ah ok, sounds good. I’ll make the change. I was wondering if we should offer an option to add `@objc` to the user, in case that’s what they want?

jrose-apple commented:
Oh, right, this is for `#selector`, not trying to satisfy another protocol. Sorry! I guess the fix-it does make sense but it's worth checking that the decl context is actually a ProtocolDecl, not just that it has a SelfProtocolDecl (because that covers extensions too).

jrose-apple commented on lib/Sema/CSApply.cpp:
```diff
@@ -4196,14 +4196,22 @@ namespace {
+          
+          tc.diagnose(E->getLoc(), diag::expr_selector_not_objc,
+                      foundDecl->getDescriptiveKind(), foundDecl->getFullName())
           .highlight(subExpr->getSourceRange());
```
Nitpick: these chaining lines are supposed to be indented. (In our usual style continuation lines are double-indented, but we get sloppy about it for diagnostics sometimes because we chain so frequently, and in this case the dots are lining up.)

theblixguy replied:
Done! 👍

Commit:
[csapply] try casting the declcontext to protocoldecl

Commit:
[csapply] fix indent

Commit:
[csapply] fix indent on return statement

Commit:
[csapply] reflow some lines

Commit:
[test] adds test for objc selector fixit

theblixguy committed:
[test] fix indentation

theblixguy committed:
[test] fix indentation

jrose-apple commented on lib/Sema/CSApply.cpp:
```diff
@@ -4198,16 +4198,16 @@ namespace {
+          // then insert the fix-it on protocol, rather than on the method in the
           // protocol declaration (not allowed).
-          auto protocolDecl = foundDecl->getDeclContext()->getSelfProtocolDecl();
+          auto protocolDecl = dyn_cast<ProtocolDecl>(foundDecl->getDeclContext()->getAsDecl());
```
Further nitpick: we've got it set up so you can cast from DeclContext directly to Decl subclasses, so the `getAsDecl()` isn't needed.

theblixguy replied:
Oh! Fixed now.

Commit:
[csapply] check if protocol contains any associated types

Commit:
[csapply] remove comment

Commit:
Merge branch 'fix/SR-9391' of https://github.com/theblixguy/swift into fix/SR-9391

Commit:
[csapply] check if protocoldecl is not null before accessing members

Commit:
[sema] add diagnostic for when selector cannot be used

Commit:
[csapply] add associatedtype check diagnostic

Commit:
[sema] update decl name type

Commit:
[csapply] minor change to reflect new diagnostic

Commit:
[test] add objc selector test (using protocol with associatedtype)

theblixguy committed:
[test] fix test file indentation

theblixguy commented on include/swift/AST/DiagnosticsSema.def:
```diff
@@ -550,6 +550,9 @@ ERROR(expr_selector_property_not_settable,none,
 ERROR(expr_selector_property_setter_inaccessible,none,
       "setter of %0 %1 is inaccessible", (DescriptiveDeclKind, DeclName))
+ERROR(expr_selector_cannot_be_used,none,
+     "cannot use %0 as a selector due to associatedtype requirement on %1",
```
Yeah, it might be tricky to check all those cases though. At the moment, it only checks for presence of `associatedtype` members, hence why the diagnostic mentions it. I guess I could reword it to something more specific, like `cannot use %0 as a selector because %1 is not exposed to Objective-C due to associatedtype requirements`?

jrose-apple replied:
That's not correct, though. Most protocols aren't exposed to Objective-C even when they can be.

theblixguy replied:
Oh, I didn't know that! How about `cannot use %0 as a selector because it is not exposed to Objective-C due to associatedtype requirement on %1`? It's what `expr_selector_not_objc` says and I have just altered it a bit to include `associatedtype`.

theblixguy replied:
Hmm, if I don't add this check, then the fixit would be applied on the protocol & then you'll get another error: `Associated type 'Foo' cannot be declared inside '@objc' protocol 'Baz'`. 

I could reword it to something like `cannot use %0 as a selector because %1 is not exposed to Objective-C` & then add a note `It is not possible to expose a protocol to Objective-C if it has associatedtype requirements, non-bridged Swift types, etc`.  

Is this better? or do you have any refinements in mind that I could make to the diagnostic?

jrose-apple replied:
Honestly, at this point I think it'd be better not to offer the fix-it at all. "cannot use %0 as a selector because protocol %1 is not exposed to Objective-C" is probably the best we can do. I already felt weird about offering a fix-it that might not actually work *and* that might not be what the user wants (such as in the original bug report, where the default implementation won't work because it can't be exposed to Objective-C).

theblixguy replied:
@jrose-apple Hmm seems like a reasonable approach. I have updated the code now!

theblixguy commented on test/FixCode/fixits-apply-objc.swift.result:
```diff
@@ -20,3 +20,18 @@ func foo1(_ an : Any) {
   return messageData as AnyObject?
 }
+
+@objc protocol SomeProtocol {
```
So you mean we shouldn't be offering a fix-it to the user i.e just a diagnostic error which mentions adding `@objc` to the protocol, but not offering any fix-it?

jrose-apple replied:
Oh, sorry. I thought this was a test of what fix-its get auto-applied by the migrator, and it isn't.

Commit:
[test] update tests

Commit:
[sema] update diagnostic message

Commit:
[csapply] don't offer a fixit if declcontext is a protocol

Commit:
[test] fix conflicts in test

theblixguy committed:
[test] update indentation

jrose-apple commented on include/swift/AST/DiagnosticsSema.def:
```diff
@@ -551,7 +551,7 @@ ERROR(expr_selector_property_not_settable,none,
       "setter of %0 %1 is inaccessible", (DescriptiveDeclKind, DeclName))
 ERROR(expr_selector_cannot_be_used,none,
-     "cannot use %0 as a selector due to associatedtype requirement on %1",
+     "cannot use %0 as a selector because protocol %1 is not exposed to Objective C",
```
Nitpick: "Objective-C", with the dash!

theblixguy replied:
Fixed!

jrose-apple commented on test/decl/protocol/protocols.swift:
```diff
@@ -513,16 +513,15 @@ extension LetThereBeCrash {
-  let _ = #selector(anotherFunction) // expected-error {{cannot use 'anotherFunction' as a selector due to associatedtype requirement on 'SomeProtocol'}}
- }
+	func someFunction() {
+		let _ = #selector(anotherFunction) // expected-error {{cannot use 'anotherFunction' as a selector because protocol 'SomeProtocol' is not exposed to Objective C}}
```
You can test that there are no fix-its attached to an `expected-*` diagnostic by adding `{{none}}` after the first `{{...}}`.

(Also nitpick: tabs snuck in here.)

theblixguy replied:
Fixed!

jrose-apple reviewed (commented):
Thanks for sticking with this one.

theblixguy committed:
[sema] rename "Objective C" to "Objective-C"

theblixguy committed:
[test] add {{none}}

theblixguy committed:
[test] update diagnostic message

theblixguy commented:
@jrose-apple can you invoke the CI?

jrose-apple commented on lib/Sema/CSApply.cpp:
```diff
@@ -4196,13 +4196,24 @@ namespace {
+         if (auto protocolDecl = dyn_cast<ProtocolDecl>(foundDecl->getDeclContext())) {
+             tc.diagnose(E->getLoc(), diag::expr_selector_cannot_be_used,
+                         foundDecl->getBaseName(), protocolDecl->getFullName());
+             return E;
```
Nitpick: extra indentation

jrose-apple commented on lib/Sema/CSApply.cpp:
```diff
@@ -4196,13 +4196,24 @@ namespace {
+         tc.diagnose(foundDecl, diag::make_decl_objc,
+                     foundDecl->getDescriptiveKind())
+            .fixItInsert(foundDecl->getAttributeInsertionLoc(false),
+                         "@objc ");
```
Wrong indentation here too. It's either 4-space (because it's a wrapped statement) or 2-space to align the dots.

jrose-apple commented:
@swift-ci Please test

theblixguy committed:
[csapply] fix indentation

theblixguy committed:
[csapply] fix indentation

theblixguy committed:
[test] add import ObjectiveC

theblixguy commented:
@jrose-apple I have fixed the indentation. The tests were failing due because of missing ObjectiveC import, so I have added that to the test file. Can you invoke the CI?

harlanhaskins commented:
@swift-ci please test

Commit:
[test] move test to selector.swift

Commit:
[test] remove newline

theblixguy commented:
Seems like adding the ObjectiveC import to that file was a bad idea. I have now moved my test to `expr/primary/selector/selector.swift`, which looks like the right place to put the tests and already contains other related `#selector` tests, so it should all be fine now 🙏  

@harlanhaskins can you invoke the CI again? 😄

harlanhaskins commented:
@theblixguy 👍

@swift-ci please test

theblixguy commented:
@harlanhaskins thanks, all tests have passed 🎉

jrose-apple commented on lib/Sema/CSApply.cpp:
```diff
@@ -4196,6 +4196,17 @@ namespace {
+        // protocol might not be the right thing to do and could lead to
+        // problems.
+        if (auto protocolDecl = dyn_cast<ProtocolDecl>(foundDecl->getDeclContext())) {
+            tc.diagnose(E->getLoc(), diag::expr_selector_cannot_be_used,
```
I'm sorry, the indentation here is *still* wrong. Two spaces for nested blocks!

theblixguy committed:
[csapply] fix indentation

theblixguy committed:
[sema] [diags] fix spacing

theblixguy commented on include/swift/AST/DiagnosticsSema.def:
```diff
@@ -550,6 +550,9 @@ ERROR(expr_selector_property_not_settable,none,
 ERROR(expr_selector_property_setter_inaccessible,none,
       "setter of %0 %1 is inaccessible", (DescriptiveDeclKind, DeclName))
+ERROR(expr_selector_cannot_be_used,none,
+     "cannot use %0 as a selector because protocol %1 is not exposed to Objective-C",
```
Done 👍

CodaFi commented:
LGTM.  Will manually squash this one 

@swift-ci please smoke test

CodaFi commented:
@swift-ci please smoke test

CodaFi commented:
@theblixguy Thanks for your patience.  Please update the JIRA ticket.
